upgraded 2103 ug from 2101 ug (#307)
AY2324S2-CS2103T-W10-1/tp: docs/UserGuide.md

**General Command Structure:**

> * Commands typically follow this format: `COMMANDWORD PREFIX_ONE/UPPER_CASE PREFIX_TWO/UPPER_CASE ....` This structure helps in **executing specific actions** within the application.

Reviewer: ```suggestion
* Commands typically follow this format: `COMMANDWORD PREFIX_ONE/PARAM_ONE PREFIX_TWO/PARAM_TWO ....` This structure helps in **executing specific actions** within the application.
```

im wondering if this is a better way bc i think UPPER_CASE might be misleading

> * `COMMANDWORD` refers to the various commands user can input. 

Reviewer: repeated?

  * Examples include `add`, `delete`, `filter`. 
  * A summary of all commands and their usages can be found in the [Commands Summary](#command-summary) table. 
* Words in `UPPER_CASE` are the **parameters to be supplied by the user**.<br>
    e.g. in `add n/NAME ....`, `NAME` is a parameter which can be used as `add n/John Doe ....`. 
  * For more details on the parameters, do refer to the [Field Constraints](#field-constraints).
* Parameters enclosed in `[]` are **optional to input**.<br>e.g. `[r/REMARK]` indicates an optional input field for that command.
* You may input the parameters **in any order** (e.g. if the command specifies `n/NAME a/ADDRESS`, `a/ADDRESS n/NAME` is also acceptable).

**Command Words:**

> * `COMMANDWORD` refers to the **base commands** that the user can input, such as `add`, `delete`, `filter`. These commands are _case-insensitive_, meaning both `add` and `ADD` will perform the addition operation.

Reviewer: repeated with above

**Prefix Usage:**

> * `PREFIX/` tags are used to **specify the type of data being entered**. For example, in `add n/John Doe p/98765432`, `n/` stands for name, and `p/` stands for phone number. Like command words, prefixes are also case-insensitive (`n/` and `N/` are treated the same).
> * `PREFIX/` refers to the prefix tag associated with each user-written input. <br>

Reviewer: ```suggestion
* `PREFIX/` are used to **specify the type of data being entered**. For example, in `add n/John Doe p/98765432`, `n/` stands for name, and `p/` stands for phone number. Like command words, prefixes are also case-insensitive (`n/` and `N/` are treated the same).
* `PREFIX/` refers to the prefix associated with each user-written input. <br>
```

could consider not using the word tag as it might get confused with our tags

also a repeat of prefixes being case insensitive

  e.g. in `add n/NAME r/REMARK`, `NAME` is the name parameter and `REMARK` is the remark parameter.

> * `PREFIX` is **case-insensitive**, i.e.`N/` and `n/` are the same and are used to indicate a `NAME` input.
> * `PREFIX` must be **preceded by a whitespace character**.

Reviewer: ```suggestion
* Prefixes are **case-insensitive**, i.e.`N/` and `n/` are the same and are used to indicate a `NAME` input.
* Prefixes must be **preceded by a whitespace character**.
```

`add n/NAME p/PHONE i/INCOME e/EMAIL a/ADDRESS f/FAMILY t/TAG h/HOUSING_TYPE [r/REMARK] [b/BIRTHDAY]`
</box>

> <box type="info" header="Notes">

Reviewer 2: I think Features section should keep/add and reference to Field Constraints

![image](https://github.com/AY2324S2-CS2103T-W10-1/tp/assets/39399745/8c00339a-ea7f-4ed8-a5ed-5947f1580973)

Some of the points should be separated, 
`PREFIX/ tags are used to specify the type of data being entered. For example, in add n/John Doe p/98765432, n/ stands for name, and p/ stands for phone number. Like command words, prefixes are also case-insensitive (n/ and N/ are treated the same).`
one point for what is prefix and another for case insensitive, they are two features all together

- You **cannot** have duplicate persons with the same name in Realodex.
    - Names are case-insensitive as described in [Field Constraints](#field-constraints).
    - If you try to add duplicate persons, you will get the error message "This client already exists in Realodex".

</box>

<box type="info" header="Notes">

  * `n/NAME`,`p/PHONE`,`i/INCOME`,`e/EMAIL`,`a/ADDRESS`,`f/FAMILY`,`t/TAG` and `h/HOUSING_TYPE` are **compulsory fields**.
    * If any of the above fields are missed out in the `add` command, you will receive an error message informing you of the compulsory fields that you missed. 
      * For example, if only `n/NAME` and `i/INCOME` are present, you will be alerted that you are missing the fields `p/PHONE`,`e/EMAIL`,`a/ADDRESS`,`f/FAMILY`,`t/TAG`and`h/HOUSINGTYPE`.

  * There can only be **one of each field** in the add command except for `t/TAG` which accepts **up to 2 tags**.
- For `t/TAG`, you may input both `t/BUYER` and/or `t/SELLER` (case insensitive).
  - If either tag is repeated more than once, for e.g. `t/BUYER t/BUYER t/BUYER`, the tag will only be recorded once and no error will be thrown.

> - `r/REMARK` and `b/BIRTHDAY` fields are optional, enclosed in `[]`. You may choose to omit them.
>   - If you include the prefix with a blank input, the birthday and remark fields will be taken as not specified.
>     - Example: `add n/John Doe p/98765432 i/20000 e/johnd@example.com a/311, Clementi Ave 2, #02-25 f/4 t/Buyer h/HDB r/ b/` will successfully add John Doe but remarks and birthday will be not specified.

Reviewer 2: Small thing

```suggestion
- `r/REMARK` and `b/BIRTHDAY` fields are optional, enclosed in `[]`. You may choose to omit them.
  - If you include the prefix with a blank input, the birthday and remark fields will be taken as not specified.
    - Example: `add n/John Doe p/98765432 i/20000 e/johnd@example.com a/311, Clementi Ave 2, #02-25 f/4 t/Buyer h/HDB r/ b/` will successfully add John Doe but remarks and birthday will be not specified.
  - You may also choose not to put the prefixes altogether. 
     -    - Example: `add n/John Doe p/98765432 i/20000 e/johnd@example.com a/311, Clementi Ave 2, #02-25 f/4 t/Buyer h/HDB` will also successfully add John Doe but remarks and birthday will be not specified.
```

Author: 👍🏻

* Deletes the client of the specified `INDEX` in Realodex.
* 💡 If you are currently filtered, the index will be based on the filtered list.
* If `INDEX` is **more than the number of clients in Realodex**, error message will be shown "The client index provided is invalid."
  to indicate that this client index does not exist.
    * This does not apply to unrealistic index values of >= `2147483648` which results in integer overflow as expected
      and will no longer be interpreted as a non-zero unsigned integer.
      Hence, error message will be shown "Index is not a non-zero unsigned integer."

</box>

<box type="wrong" header="Error">

- If `INDEX` is a **non-zero unsigned integer**, error message will be shown "Index is not a non-zero unsigned integer."
- If there is no client with the specified `INDEX` in the current list, an error message will be shown: "The client index provided is invalid".
- If neither index nor name is provided `delete` will show an error message "Please provide either an index or a name."
- If both an index and name is provided `delete INDEX n/NAME` will show an error message "Please provide either an index or a name, not both."

> - If both an index and name is provided `delete n/NAME INDEX ` will show an error message "The client name provided is invalid" as INDEX is considered part of the NAME."

Reviewer: ```suggestion
- If both an index and name is provided `delete n/NAME INDEX ` will show an error message "The client name provided is invalid" as INDEX is considered part of the NAME.
```

`delete INDEX`
</box>
<box type="info" header="Notes">

* Deletes the client of the specified `INDEX` in Realodex.
* 💡 If you are currently filtered, the index will be based on the filtered list.
* If `INDEX` is **more than the number of clients in Realodex**, error message will be shown "The client index provided is invalid."
  to indicate that this client index does not exist.
    * This does not apply to unrealistic index values of >= `2147483648` which results in integer overflow as expected
      and will no longer be interpreted as a non-zero unsigned integer.

>       Hence, below error message applies.

Reviewer 2: ```suggestion
      Hence, error message will be shown "Index is not a non-zero unsigned integer."
```

Reviewer 2: Since separate pop out boxes

Author: 👍🏻

</box>

<box type="wrong" header="Error">

- If `INDEX` is a **non-zero unsigned integer**, error message will be shown "Index is not a non-zero unsigned integer."

> - If there is no client with the specified INDEX in the current list, an error message will be shown: "The client index provided is invalid".

Reviewer 2: ```suggestion
- If there is no client with the specified `INDEX` in the current list, an error message will be shown: "The client index provided is invalid".
```

Author: 👍🏻

- If you currently have a filtered list after `filter` operations,
  the index will be based on the filtered list.
- It is optional to edit any field (i.e, you can choose to edit any combination of fields so long there is **at least 1**).
- All fields must follow the respective [Field Constraints](#field-constraints).

</box>

<box type="warning" header="Caution">

- The current information will be overwritten with the input provided. 
- When editing the `TAG`, all existing tags will be overwritten with the new tag(s) provided. If you want to edit the client to be both a buyer and seller, include both tags i.e. `t/Buyer t/Seller`.
</box>
<box type="wrong" header="Error">

- If `INDEX` is **more than the number of clients in Realodex**, error message will be shown "The client index provided is invalid."
  to indicate that this client index does not exist.
    * This does not apply to unrealistic index values of >= `2147483648` which results in integer overflow as expected
      and will no longer be interpreted as a non-zero unsigned integer.
      Hence, below error message applies.
</box>
  
**Example:**

> - edit 1 p/999 will overwrite the 1st client's phone number to "999".

Reviewer: ```suggestion
- `edit 1 p/999` will overwrite the 1st client's phone number to "999".
```

<box type="info" header="Notes">

- If `INDEX` is `3`, the 3rd client's information will be edited.
- If `INDEX` is a **non-zero unsigned integer**, error message will be shown "Invalid command format..."
- If you currently have a filtered list after `filter` operations,
  the index will be based on the filtered list.
- It is optional to edit any field (i.e, you can choose to edit any combination of fields so long there is **at least 1**).
- All fields must follow the respective [Field Constraints](#field-constraints).

</box>

<box type="warning" header="Caution">

- The current information will be overwritten with the input provided. 

> - When editing the `TAG`, all existing tags will be overwritten with the new tag(s) provided. If you want to edit the client to be both a buyer and seller, include both tags i.e. `t/Buyer t/Seller`.

Reviewer 2: LGTM

This **includes** clients who may have **both "Buyer" and "Seller" tags.**
- **Filtering by Multiple Tags**: Realodex supports filtering by **multiple tags**.
Entering `filter t/Buyer t/Seller` will show only clients who are **tagged as both "Buyer" and "Seller"**.
- **Comprehensive searching**, returning all persons' with the specified tag(s).
  </box>

> 
> <box type="warning" header="Caution">
> 
> - Tag input should be **valid** and **non-empty** - "Buyer" or "Seller".
>     - `filter t/buyer` matches person with tag "Buyer".
> - **Inclusive matching** of persons with multiple tags, as long as they possess the
>   tag(s) specified in the input.
>     - `filter t/buyer` matches person with tags "Buyer" and "Seller".
> 
> </box>
> 

Reviewer 2: ```suggestion

<box type="warning" header="Caution">
- Tag input should be **valid** and **non-empty** , that is should only be either "Buyer" or "Seller" (case-insensitive).
    - `filter t/buyer` matches person with tag "Buyer".
- **Inclusive matching** of persons with multiple tags, as long as they possess the
  tag(s) specified in the input.
    - `filter t/buyer` matches person with tags "Buyer" and "Seller".

</box>

```

Author: 🦖

<box type="info" header="Notes">

- **Partial fragments** of remarks will still be matched.
    - `filter r/hand` matches person with remark "handsome".
- **Comprehensive searching**, returning all persons' names containing the keyword .
    - `filter r/love` returns persons with remarks "loves to travel", "has a lovely dog".
</box>

<box type="warning" header="Caution">

- Keyphrase input should be **non-empty**. This is an intentional design choice to ensure that the command is used for targeted searches, preventing the potential misinterpretation of an empty keyphrase as a request to list all clients.
- The remarks for the `filter r/` command **must not contain any other prefixes** to prevent parsing errors.
</box>

<box type="wrong" header="Error">

- The command `filter r/ my tag is t/buyer` would cause an error because the system interprets `t/` as the start of a new prefix.
- To avoid this, ensure that the remark does not contain any spaces followed by slashes that could be misconstrued as additional prefixes.
</box>

> 

Reviewer 2: 👍

<box type="info" header="Notes">

- Month input should be a valid month in **3-letter abbrieviation (MMM)** format and **non-empty**.
  - E.g., "Jan" for January, "Feb" for February 
  - Filtering by month "September" should be `filter b/Sep`
    - `filter b/SEP` matches all people with Birthday in September.
- Month input will also accept extraneous input, i.e. if the first 3 characters are any valid month, it will be valid. 
  - E.g., This forgiving behavior allows you to type in `filter b/AprMogger` and Realodex will interpret the month as April!
- **Comprehensive searching**, returning all persons with birthdays in the specified month.
    - `filter b/Jan` returns all persons with birthday in January.
</box>

<box type="warning" header="Caution">

> - Persons who do not have a specified birthday will **not be included** in the search results.

Reviewer 2: 👍

<box type="info" header="Notes">

* The current date is based on the local system's time. 
* If their birthday has already passed, the calculation is based on the number of days until their next birthday next year.
</box>

<box type="warning" header="Warning">

* If the list presented is currently a filtered list after using [filter](#filtering-clients-filter), sort will work on the new filtered list.
* If a birthday falls on February 29th (leap day), the day calculation is based on March 1st if the year does not have a leap date as realistically, most would still celebrate every year.

> </box>

Reviewer 2: I think can move these warnings to notes as they are not exactly cautionary, small thing LOL

Author: 🌮

> | `i/Income`                                                            | <ul><li>**Integer** and should be **at least 0**. </li></ul>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | <ul><li>✅ <code>i/10000</code></li><li>❌ <code>i/10,000</code></li><li>❌ <code>i/-1</code></li></ul>                  |

Reviewer 2: Actually should we add a constraint for income smth like for phone, if they want to put commas or spaces, a work around is to avoid the use of symbols, like they may think `1,000,000` is a valid integer but we wont allow that hence FF bla bla, so maybe can hedge first

Author: actl this is there already tho

Reviewer 2: No to justify why we dont allow, give them a workaround

| `e/Email`                                                             | <ul><li>Emails should be of the format local-part@domain.</li><li>The local-part should only contain alphanumeric characters and these special characters: +_.- (excluding parentheses).</li><li>The local-part may not start or end with any special characters.</li><li>This is followed by a '@' and then a domain name, made up of domain labels separated by periods.</li><li>The domain name must end with a domain label at least 2 characters long.</li><li>Each domain label must start and end with alphanumeric characters.</li><li>Each domain label may consist of alphanumeric characters, separated only by hyphens, if any.</li><li>Top-level domain (TLD) such as .com are not compulsory, as according to Internet Protocol Standards (<a href="https://datatracker.ietf.org/doc/html/rfc5322#section-3.4.1">RFC 5322 Section 3.4.1</a>).</li></ul> | <ul><li>✅ <code>e/admin@realodex</code></li><li>❌ <code>e/hello@x_mail</code></li></ul>                               |

> | `a/Address`                                                           | <ul><li>Must not include other command prefixes (`a/`,`b/`,`e/`,`f/`,`h/`,`i/`,`n/`,`p/`,`r/`,`t/`) to prevent parsing errors. For instance, `a/lemontree street t/1` may cause the command to fail, as the system will interpret `t/` as an unintended tag prefix.</li></ul>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | <ul><li>✅ <code>a/6 College Avenue West</code></li>                                                                   |

Reviewer 2: ```suggestion
| `a/Address`                                                           | <ul><li>Must not include other command prefixes with spaces before (`a/`,`b/`,`e/`,`f/`,`h/`,`i/`,`n/`,`p/`,`r/`,`t/`) to prevent parsing errors. For instance, `a/lemontree street t/1` may cause the command to fail, as the system will interpret `t/` as an unintended tag prefix.</li></ul>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | <ul><li>✅ <code>a/6 College Avenue West</code></li>                                                                   |
```

Reviewer 2: `with spaces before ` basically

Author: 🇹🇼

| `h/HousingType`                                                       | <ul><li>Must be one of the following: **"HDB", "CONDOMINIUM", "LANDED PROPERTY", "GOOD CLASS BUNGALOW"** (case-insensitive).</li><li> Only **one housing type** is allowed.   </li></ul>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | <ul><li>✅ <code>h/HDB</code></li></ul><ul><li>❌ <code>h/big house</code></li></ul>                                    |

> | `r/Remark`                                                            | <ul><li>**Can be empty** if remark is not specified.</li><li>Must not include other command prefixes (`a/`,`b/`,`e/`,`f/`,`h/`,`i/`,`n/`,`p/`,`r/`,`t/`) to prevent parsing errors. For instance, `a/lemontree street t/1` may cause the command to fail, as the system will interpret `t/` as an unintended tag prefix.</li></ul>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | <ul><li>✅ <code>r/Likes to eat cake</code></li></ul>                                                                  |

Reviewer 2: ```suggestion
| `r/Remark`                                                            | <ul><li>**Can be empty** if remark is not specified.</li><li>Must not include other command prefixes with spaces before (`a/`,`b/`,`e/`,`f/`,`h/`,`i/`,`n/`,`p/`,`r/`,`t/`) to prevent parsing errors. For instance, `a/lemontree street t/1` may cause the command to fail, as the system will interpret `t/` as an unintended tag prefix.</li></ul>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | <ul><li>✅ <code>r/Likes to eat cake</code></li></ul>                                                                  |
```

Reviewer 2: Similar to address

| `b/Birthday`                                                          | <ul><li>BIRTHDAY should be in the form "DDMMMYYYY", and can be empty if the birthday is not specified. <li>The date must not be in the future.</li><li>The date must exist in the Gregorian calendar. (<code>b/29Feb2023</code> is not allowed as it is not a valid day to begin with)</li><li>The day "DD" must be numeric. For 1st-9th day of the month, the 0 need not be present. </li><li>The month "MMM" refers to the first 3 letters of the month (case-insensitive)</li><li>The year "YYYY" must be in full and greater than or equal to 1000.</li></ul>                                                                                                                                                                                                                                                                                                     | <ul><li>✅ <code>b/2Feb2002</code></li><li>✅ <code>b/23Apr1972</code></li><li>❌ <code>b/23Apr72</code></li></ul> </ul> |

> 

Reviewer 2: 😳

> 

Reviewer 2: I cant add comment below this but anyways for

Command Summary
**Help (individual commands)** i think best to link to the actual help command as they may again raise the flaw of why some commands are not supported

Author: actually the only one that is not covered is exit

Author: i made the change:

* For the full list of commands, you can use the general [help command](#help-help)
